nvflare/job_config/base_script_runner.py

```diff
@@ -0,0 +1,242 @@
+from nvflare.app_common.abstract.launcher import Launcher
+from nvflare.app_common.executors.client_api_launcher_executor import ClientAPILauncherExecutor
+from nvflare.app_common.executors.in_process_client_api_executor import InProcessClientAPIExecutor
+from nvflare.app_common.launchers.subprocess_launcher import SubprocessLauncher
+from nvflare.app_common.widgets.external_configurator import ExternalConfigurator
+from nvflare.app_common.widgets.metric_relay import MetricRelay
+from nvflare.client.config import ExchangeFormat, TransferType
+from nvflare.fuel.utils.import_utils import optional_import
+from nvflare.fuel.utils.pipe.cell_pipe import CellPipe
+from nvflare.fuel.utils.pipe.pipe import Pipe
+
+from .api import FedJob, validate_object_for_job
+
+
+class FrameworkType:
+    RAW = "raw"
+    NUMPY = "numpy"
+    PYTORCH = "pytorch"
+    TENSORFLOW = "tensorflow"
+
+
+class BaseScriptRunner:
+    def __init__(
+        self,
+        script: str,
+        script_args: str = "",
+        launch_external_process: bool = False,
+        command: str = "python3 -u",
+        framework: FrameworkType = FrameworkType.PYTORCH,
+        params_transfer_type: str = TransferType.FULL,
+        task_pipe: Optional[Pipe] = None,
+        executor: Union[ClientAPILauncherExecutor, InProcessClientAPIExecutor, None] = None,
+        launcher: Optional[Launcher] = None,
+        metric_pipe: Optional[Pipe] = None,
+        metric_relay: Optional[MetricRelay] = None,
+    ):
+        """BaseScriptRunner is used with FedJob API to run or launch a script.
+
+        If executor is not provided,
+            `launch_external_process=False` uses InProcessClientAPIExecutor.
+            `launch_external_process=True` uses ClientAPILauncherExecutor.
+        else the provided executor will be used.
+
+        If some components are passed in, it is user's responsibility to make sure they are consistent with each other.
+        For example, if user provide a task_pipe, the default task_pipe_id is "task_pipe",
+        please make sure the ClientAPILauncherExecutor is created using the matching pipe_id.
+
+        Args:
+            script (str): Script to run. For in-process must be a python script path. For ex-process can be any script support by `command`.
+            script_args (str): Optional arguments for script (appended to script).
+            launch_external_process (bool): Whether to launch the script in external process. Defaults to False.
+            command (str): If launch_external_process=True, command to run script (preprended to script). Defaults to "python3".
+            framework (str): Framework type to connfigure converter and params exchange formats. Defaults to FrameworkType.PYTORCH.
+            params_transfer_type (str): How to transfer the parameters. FULL means the whole model parameters are sent.
+                DIFF means that only the difference is sent. Defaults to TransferType.FULL.
```

Comment:
Please add description for other newly added args.

Reply:
thanks, added.